occurrences/tests/test_api.py

```diff
@@ -0,0 +1,341 @@
+        "pEventId": "",
+    }
+}
+
+UPDATE_OCCURRENCE_MUTATION = """
+mutation updateOccurrence($input: UpdateOccurrenceMutationInput!){
+  updateOccurrence(input: $input){
+    occurrence{
+      minGroupSize
+      maxGroupSize
+      organisation{
+        name
+      }
+      contactPersons{
+        edges {
+          node {
+            name
+          }
+        }
+      }
+      startTime
+      endTime
+      pEvent{
+        duration
+        neededOccurrences
+        enrolmentEnd
+        enrolmentStart
+        linkedEventId
+      }
+    }
+  }
+}
+"""
+
+UPDATE_OCCURRENCE_VARIABLES = {
+    "input": {
+        "placeId": "place_id",
+        "minGroupSize": 10,
+        "maxGroupSize": 20,
+        "startTime": "2020-05-05T00:00:00+00",
+        "endTime": "2020-05-05T00:00:00+00",
+        "organisationId": "",
+        "contactPersons": [
+            {"id": "", "name": "New name", "emailAddress": "newname@email.address"},
+        ],
+        "pEventId": "",
+    }
+}
+
+DELETE_OCCURRENCE_MUTATION = """
+mutation DeleteOccurrence($input: DeleteOccurrenceMutationInput!) {
+  deleteOccurrence(input: $input) {
+    __typename
+  }
+}
+"""
+
+
+def test_study_groups_query(snapshot, study_group, api_client):
+    executed = api_client.execute(STUDY_GROUPS_QUERY)
+    snapshot.assert_match(executed)
+
+
+def test_study_group_query(snapshot, study_group, api_client):
+    executed = api_client.execute(
+        STUDY_GROUP_QUERY,
+        variables={"id": to_global_id("StudyGroupNode", study_group.id)},
+    )
+    snapshot.assert_match(executed)
+
+
+def test_occurrences_query(snapshot, occurrence, api_client):
+    executed = api_client.execute(OCCURRENCES_QUERY)
+    snapshot.assert_match(executed)
+
+
+def test_occurrence_query(snapshot, occurrence, api_client):
+    executed = api_client.execute(
+        OCCURRENCES_QUERY,
+        variables={"id": to_global_id("OccurrenceNode", occurrence.id)},
+    )
+    snapshot.assert_match(executed)
+
+
+def test_add_occurrence_unauthenticated(api_client, user_api_client):
+    executed = api_client.execute(
+        ADD_OCCURRENCE_MUTATION, variables=ADD_OCCURRENCE_VARIABLES
+    )
+    assert_permission_denied(executed)
+
+    executed = user_api_client.execute(
+        ADD_OCCURRENCE_MUTATION, variables=ADD_OCCURRENCE_VARIABLES
+    )
+    assert_permission_denied(executed)
+
+
+def test_add_occurrence(snapshot, staff_api_client, organisation, person, p_event):
+    variables = deepcopy(ADD_OCCURRENCE_VARIABLES)
+    variables["input"]["organisationId"] = to_global_id(
+        "OrganisationNode", organisation.id
+    )
+    variables["input"]["pEventId"] = to_global_id(
+        "PalvelutarjotinEventNode", p_event.id
+    )
+    # Add one more pre-made contact person to the variable
+    variables["input"]["contactPersons"].append(
+        {
+            "id": to_global_id("PersonNode", person.id),
+            "emailAddress": person.email_address,
+            "name": person.name,
```

Comment:
Well yeah, one problem is here the input also allows adding new person. So both adding new person or adding existed person are sharing the same input template, which mean `name` and `email` fields are always required. It's actually not that bad since in the UI there is a select-dropdown to select person so the person data can be easily collected through the form. This recalls me of the same thing in kukkuu-admin translatable fields. Sometimes it's better to just send full data :D Anw good point, I'll note it.